packages/dds/tree/src/core/tree/detachedFieldIndex.ts

```diff
@@ -145,24 +165,86 @@ export class DetachedFieldIndex {
-						this.detachedNodeToField,
-						nodeId.major,
-						nodeId.minor + i,
-					) === undefined,
-					0x7ce /* Detached node ID already exists in index */,
-				);
-				setInNestedMap(this.detachedNodeToField, nodeId.major, nodeId.minor + i, root + i);
-			}
+			this.setDetachedNodeRange(nodeId, count, root);
 		}
 		return root;
 	}
 
+	private setDetachedNodeRange(
+		nodeId: Delta.DetachedNodeId,
+		count: number,
+		root: ForestRootId,
+	): void {
+		if (!this.detachedNodeRangeMap.has(nodeId.major)) {
+			this.detachedNodeRangeMap.set(nodeId.major, []);
+		}
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		const innerRangeMap = this.detachedNodeRangeMap.get(nodeId.major)!;
+		assert(
+			getFirstEntryFromRangeMap(innerRangeMap, nodeId.minor, count) === undefined,
+			"The detached node range already exists in the map",
+		);
+		setInRangeMap(innerRangeMap, nodeId.minor, count, root);
+	}
+
+	/**
+	 * Given a query range, find and group all detached nodes within the range according to the following rules:
+	 *
+	 * 1. If a range entry overlaps with the query range, store the overlapping portion as a new range in the result.
+	 * 2. If there is an "empty" space between two consecutive range entries, store the range between the end
+	 * of the first entry and the start of the second entry as a new range with an undefined value.
+	 *
+	 * The function only returns the length and value (root) of the range entry. Since consecutive ranges without gaps
+	 * are always returned, there's no need to include the start points of the ranges.
+	 */
+	public getAllDetachedNodeRanges(
+		nodeId: Delta.DetachedNodeId,
+		count: number,
+	): { length: number; root?: ForestRootId }[] {
+		const innerRangeMap = this.detachedNodeRangeMap.get(nodeId.major);
+		if (!innerRangeMap) {
+			return [{ length: count }];
+		}
+
+		const results = [];
+		let currentPos = nodeId.minor;
+		const validRanges = getAllValidEntriesFromMap(innerRangeMap, nodeId.minor, count);
+		for (const range of validRanges) {
+			if (currentPos < range.start) {
+				results.push({
+					length: range.start - currentPos,
+				});
+			}
+			results.push({ length: range.length, root: range.value });
+			currentPos = range.start + range.length;
+		}
+
+		if (currentPos < nodeId.minor + count) {
+			results.push({
+				length: nodeId.minor + count - currentPos,
+			});
+		}
+		return results;
+	}
+
 	public encode(): JsonCompatibleReadOnly {
+		// Since currently the codec only accepts the original NestedMap format data, so break the
+		// detached node ranges atomically into single nodeIds, and build a NestedMap from them.
+		const detachedNodeToField: NestedMap<Major, Minor, ForestRootId> = new Map();
+		for (const [major, rangeMap] of this.detachedNodeRangeMap) {
+			const innerMap = new Map();
+			for (const rangeEntry of rangeMap) {
+				for (let offset = 0; offset < rangeEntry.length; offset++) {
+					innerMap.set(rangeEntry.start + offset, rangeEntry.value);
```

Comment:
There is a 1:1 mapping between `DetachedNodeId`s and `ForestRootId`s.
```suggestion
					innerMap.set(rangeEntry.start + offset, brand(rangeEntry.value + offset));
```